### DISCUSSION TITLE
Translation scalability issue

### DISCUSSION THREAD

**Discussion by ShadowyNeoBTC** [sovenor/bitcoin-rocks#3] [Q&A]
I was reading the guide for translating the content of the website in new languages.

I think that the proposed solution in CONTRIBUTING.md is far from ideal. The problem is that all HTML files will be duplicated and a modification in one file will need to be replicated in files for every language.

This quickly leads to scalability issues.

Using [jquery.i18n](https://github.com/wikimedia/jquery.i18n) would be easier and much better for scalability and maintainability.

**Comment by sovenor** [comment]
Thanks for pointing this out! I looked at the jquery.i18n repo and it looks like it could be a perfect fit. I have a question just to make sure I am implementing it in the best way:

The number of key-value pairs in each language.json file will be quite large due to the number of pages on the website and the many different text sections, but I think that's ok. I believe you're right that this will still be a much more scalable way to translate the site into multiple languages.

Then instead of having to fork the entire repo and translate the individual HTML pages, someone can simply duplicate the "en.json" file, rename it to their language, and then translate the content in that single file.

Is this how you would recommend doing it?

**Reply by ShadowyNeoBTC** [reply]
Yeah that's exactly it ! You will indeed have a large number of translation strings but that's inevitable. On the other hand, HTML files won't be duplicated.

**Reply by sovenor** [reply]
Ok great, I will work on implementing this. It seems like a more ideal tradeoff balance than the current setup. I'll keep this discussion open until my work is complete.

**Reply by sovenor** [reply]
@L01C-CODES are you able to help me by building a simple example page with all the required jquery.i18n dependencies and the associated en.json file?

It doesn't need to be big, it just needs to work so I can build on it from there. I have tried for a couple hours to get an example page working with no luck. I'm sure I must be missing something simple but I can't figure it out. Once I have an example page working I can make all the translation strings and get the entire website ready for translations.

**Reply by ShadowyNeoBTC** [reply]
Yep, I'll work on that

**Reply by sovenor** [reply]
Amazing, thank you!

**Reply by ShadowyNeoBTC** [reply]
Quick update, I've got it working. I just need to add a language switcher and I'll push to a branch

**Reply by OsodracPT** [reply]
If you need some inspiration you have the nostr how project.  [https://github.com/erskingardner/nostr-how](url)

They have also used i18n.

**Reply by sovenor** [reply]
@OsodracPT thanks for this! Do you know how they are handling the browser redirect to /en or /de or other languages? I understand they are all viewing the same HTML file, but it's nice to have the language directory present in the URL. It will be better for SEO than having all the translations accessible from the same root domain.

Please let me know if you know how they're accomplishing this.

**Reply by sovenor** [reply]
@OsodracPT nevermind, it seems they are actually using i18n for UI element translations and separate directories for individual pages. The plan for bitcoin.rocks is to use i18n for everything so it can all pull from a single HTML page. There will be a lot of localization strings in each json file but then if I make a change to the HTML page I don't have to make that same change in each local directory.

**Comment by ShadowyNeoBTC** [comment]
Here is the code with dynamic translation. 
https://github.com/L01C-CODES/bitcoin-rocks

I didn't style the language switcher for sake of making this quick.

Let me know if you have questions.

**Reply by ShadowyNeoBTC** [reply]
@sovenor if that works, I can submit a pull request

**Reply by sovenor** [reply]
Thanks, I will review ASAP (sometime this week)

**Reply by sovenor** [reply]
@L01C-CODES I am reviewing the code now and the changes look good! However, the repo doesn't include any of the i18n files. It seems the folder at en/jquery/jquery.18n is empty. Can you add the files and push the change so I can test it?

**Reply by sovenor** [reply]
I added the required i18n files from their repo and it's working! Amazing work, thank you so much!

Feel free to add those files and then submit a pull request. Once that's done I'll begin working on all the translation strings. Very exciting!

**Reply by ShadowyNeoBTC** [reply]
<img width="500" alt="image" src="https://github.com/sovenor/bitcoin-rocks/assets/134526159/52070806-ca31-4997-9b63-e2b58c182f0f">

Forgot to mention, it is because it's a submodule. You need to update it.

**Reply by sovenor** [reply]
Ah I see, that makes sense. You can go ahead and submit a pull request and I'll approve it.

**Comment by ObjectifMoon21** [comment]
I'm a bit lost in github, i can learn but there may be other ways of offering translations that would be easier for everyone.
 
I've used Localazy to translate an app. This program also works for translating websites, everyone can see the progress of a translation and participate or proofread. 

It's very easy to use for non programmers, translators has just to translate the words or sentences and all the part of pushing the translation in the coding files take place in the background.

https://localazy.com/term/website-content-translation
https://github.com/localazy

**Reply by sovenor** [reply]
Thanks for this suggestion. After reviewing it and the i18n suggestions above, I think i18n makes the most sense. You will still be able to translate a single file in your browser. The file will look like this: https://github.com/L01C-CODES/bitcoin-rocks/blob/main/en/i18n/en.json

All you have to do is translate the text on the right side of the pairs in the file. When it's ready, I'll publish a short guide that makes it easy to do from your browser even if you are not used to GitHub.

Please don't begin translating yet. I'll make an announcement when it's ready for translations. Thanks!

**Comment by ShadowyNeoBTC** [comment]
@sovenor saw you just merged. Just keep in mind that the language selector still needs to be styled.

**Reply by sovenor** [reply]
Thanks for the reminder. I am not planning on pushing the update to the actual web server until all of the localization strings are added, so no worries there. I'll be sure to style the language selector as part of the commit.

**Comment by sovenor** [comment]
@L01C-CODES could you please check [the i18n/en.json file in my latest commit](https://github.com/sovenor/bitcoin-rocks/commit/bf161d812db719d0a3771c595bf8c1fde491b2ce#diff-c3ec371971c6de4e02f589b8941c4e04a75d7ec9d9149b618bc43a6999a4a628) when you have a moment?

Do you think this is the right way to go about naming all these unique translation strings in the i18n/en.json file? I would love your feedback before I work on the other pages. See 404.html or bank-runs.html for corresponding translation string locations.

**Reply by ShadowyNeoBTC** [reply]
I think that makes sense yes. One thing I was wondering and that might be worth testing is if you can embedded translation strings. Doing that would allows us to split by page so that it's easier to navigate the JSON file.

So something like:

```
{
  "home": {
    "home_btn_you": "YOU"
  }
}
```

Which would be accessible by `home.home_btn_you`

**Reply by sovenor** [reply]
Great idea! I got this working but separated everything out into individual files to make it cleaner. Check my [latest commit](https://github.com/sovenor/bitcoin-rocks/commit/1796635c3f4d1c63640904b4b6f5906e82a224fa) for details.

When you are on index.html it will automatically pull in the i18n/en/index_en.json file for translations.
When you are on 404.html it will automatically pull in the i18n/en/404_en.json file, etc.

This should be a much more organized approach and make it easier for translators to just focus on 1 page at a time.

I couldn't figure out how to access translation strings from another .json file unless that file is explicitly loaded on the page, so for now I just have common_en.json for common translation strings that need to be referenced on multiple pages.

I'll work on more pages today!

**Reply by ShadowyNeoBTC** [reply]
I like that a lot ! I think one translation file per page makes sense

**Reply by sovenor** [reply]
Ok great, I'll keep cranking away!

**Comment by sovenor** [comment]
Update: I am still working on finishing all of these translation strings. I just have the business pages left and I'll be done.

I'm unsure of the best way to handle SEO text like the page title, meta description, etc.

@L01C-CODES do you have any suggestions for how to handle this? Is it ok for it to just be translation strings or is that going to nuke SEO?

**Reply by ShadowyNeoBTC** [reply]
Hmm I must admit I'm not sure what is the best way to proceed here.

**Reply by sovenor** [reply]
Ok I will do some more research and see if I can find a good solution.

**Reply by sovenor** [reply]
Here are some things I've learned:

- I can dynamically change the HTML "lang" attribute at the top of the page depending on which language the user selects in the language selector dropdown. This should help search engines know which content is related to which language.
- I can add a canonical tag to the top of each page. Most search engines will just index the default English language which is fine since the URLs will still be English words in all versions.
- It can be ok to have multiple hreflang tags at the top of each page to indicate the available languages. When a new language is added, I'll add a new hreflang tag to each page (easy with find and replace).
- I can specify language-specific meta information like page descriptions if necessary.
- Having dynamic jquery.i18n page titles may affect SEO but will allow the entire contents of the page to be translated. This should result in a better UX for everyone.

Ultimately I am ok with taking the SEO hit (if it happens) since it's such a small percentage of total traffic. Most traffic comes from people visiting the site directly or scanning the QR codes on our stickers, and I expect that to be the same in other languages too.

I think the tradeoffs here make sense to enable this content to be translated into more languages and reach more people. Onward!

**Comment by sovenor** [comment]
Update: translation strings v1 are now live on the website!

If you have a moment, please poke around [the website](https://bitcoin.rocks/) and let me know if you see any obvious bugs or suggestions. It seems everything is working from my end so far.

**Reply by ShadowyNeoBTC** [reply]
Checked most of the pages and they seem good.

One thing I noted though is that some images (linked below) that contain text could probably be done using CSS. Especially now that you will be relying on other people to translate the website, it would be a good thing that the least text possible lives inside images.

![image](https://github.com/sovenor/bitcoin-rocks/assets/134526159/5c0f3164-4f9e-4e1c-b238-dfa6b28691aa)

**Reply by ShadowyNeoBTC** [reply]
Did something really quick here but it gives an idea of what can be done with the right font and a better image for the checkmark.
![image](https://github.com/sovenor/bitcoin-rocks/assets/134526159/826ea8f4-ab30-45d2-a50d-77a07661c445)

**Reply by sovenor** [reply]
Great idea, thank you! I will add that to my list.

**Comment by sovenor** [comment]
Version 1 of the [Translation Guide](https://github.com/sovenor/bitcoin-rocks/blob/main/CONTRIBUTING.md) is now live!

This should allow anyone to start translating content into their preferred language even if they are not familiar with GitHub. The language selector dropdown on the website has an option to "Add language" which links to the guide.

@ObjectifMoon21 could you follow the steps in the [Translation Guide](https://github.com/sovenor/bitcoin-rocks/blob/main/CONTRIBUTING.md) and let me know if you run into any issues or are unsure of what to do at any point?

@OsodracPT it would be great if you could do the same and start on the Polish translations again.

@L01C-CODES if you have time to review the guide and let me know if you see any errors or anything I could change to make it easier to understand, that would be super helpful.

Thank you all!

**Reply by ObjectifMoon21** [reply]
Great @sovenor ! Thanks for the translation guide, it's clear for me now.
I've started to translate some json files in French. 
Would you prefer that i send the pull requests after some files translated or at the end of all of them?

**Reply by sovenor** [reply]
Awesome @ObjectifMoon21! Please be sure to [make a new discussion here](https://github.com/sovenor/bitcoin-rocks/discussions/categories/translations) titled "French" so others can join in if they'd like.

As long as you are the only one translating into French, feel free to wait on the pull request until the end. Thank you so much for your time on this!

**Reply by ShadowyNeoBTC** [reply]
@sovenor I think the guide is pretty good. At first I thought it might be a bit muhc guiding the user towards downloading and using the GitHub Desktop app, but after consideration I think it's okay because it allows non-technical people to easily contribute to the translation of the website without having to figure out how git works.

**Comment by sovenor** [comment]
Closing this discussion since translations are now scalable. I have added the CSS graphics suggestion by @L01C-CODES in a new issue #26 and will work on that ASAP.
